오픈 마켓 [STEP 1] Taeangel, Malrang

malrang-malrang opened the pull request:
@stevenkim18 안녕하세요 스티븐!

PR 날짜를 넉넉히 잡았다고 생각했는데....요...
프로젝트에서 필요한 개념들을 이해하는데 시간도 오래걸리고 너무어려웠습니다 🥲
좀더 공부가 필요할것같습니다!
공부하면 좋을 개념들 키워드를 많이 알려주시면 감사하겠습니다! 🤭
STEP1 PR 잘부탁드리겠습니다! 🥳

## 조언을 받고싶은점
### 1. MockURLSessionDataTask init 관련 Error
```swift
class MockURLSessionDataTask: URLSessionDataTask {
    private let closure: () -> Void

    init(closure: @escaping () -> Void) {
        self.closure = closure
    }

    override func resume() {
        closure()
    }
}
```
>![](https://i.imgur.com/TqXVi4m.png)
>
> 위의 사진과 같은 에러가 나오는데 어떤방법 키워드를 공부해야 해결할수있을지 궁금합니다!

### 2. 테스트를 하기위한 Mock 객체에서의 강제언래핑
```swift
func dataTask(
        with urlRequest: URLRequest,
        completionHandler: @escaping DataTaskCompletionHandler
    ) -> URLSessionDataTask {
        let successResponse = HTTPURLResponse(
            url: urlRequest.url!,
            statusCode: 200, httpVersion: "",
            headerFields: nil
        )

```
테스트를 위한 목객체를 만들었는데 목객체에도 강제언래핑을 지양해야하는지 궁금합니다!

malrang-malrang committed:
📝[docs]: STEP1 README Update

Taeangel committed:
➕[add]: SwiftLint 추가

malrang-malrang committed:
🪛[chore]: gitignore pods 주석제거

Taeangel committed:
✨[feat]:  Products 구현

malrang-malrang committed:
✨[feat]:  ProductCatalog 구현

Taeangel committed:
✨[feat]: Decodable Extension parse메서드 구현

malrang-malrang committed:
♻️[refactor]: Decodable parse() 메서드 프로토콜로 변경

* Catalog 프로토콜 에서 수행하도록 변경

Taeangel committed:
♻️[refactor]: protocol에서 ParseManager로 parse 변경

Taeangel committed:
✨[feat]: test코드 작성

malrang-malrang committed:
♻️[refactor]: ParseManager parse() 메서드 이동

* Decodable Extension 에서 parse() 메서드 수행하도록 변경

malrang-malrang committed:
♻️[refactor]: ProductCatalogTests 테스트 코드 수정

Taeangel committed:
✨[feat]: URLSessionProvider 구현

Taeangel committed:
✨[feat]: NetworkError 구현

malrang-malrang committed:
✨[feat]:  API 파일추가 및 구현

malrang-malrang committed:
✨[feat]:  URLSessionProtocol 파일추가 및 구현

* dataTask() 메서드 필수 구현하도록 명시
* URLSession 이 URLSessionProtocol 을 채택 하도록함

Taeangel committed:
♻️[refactor]: URLSessionProvider에서 session 변경
session의 URLSession 타입에서 URLSessionProtocol으로 변경

Taeangel committed:
✨[feat]: URLSession test를 위한 MockData 구현

malrang-malrang committed:
✨[feat]: URLSessionDataTaskMock 구현

malrang-malrang committed:
✨[feat]: URLSession test 를 위한 MockURLSession 구현

Taeangel committed:
✅ [test]: MockURLSession의 테스트코드작성

Taeangel committed:
✏️[correct]: MockURLSessionDataTask네이밍 변경

Taeangel committed:
🪛[chore]: Model의 프로토콜을 Codable로변경

malrang-malrang committed:
✅ [test]: MockURLSession의 테스트코드 추가

malrang-malrang committed:
♻️[refactor]: URLSessionProvider getData() 메서드 수정

* data, error 의  early return 구문 분리

Taeangel committed:
✨[feat]: URLSessionProvider fetchData 메서드 구현

Taeangel committed:
🔧[file]: test코드 파일 삭제및 생성

Taeangel committed:
✅ [test]: HealthChecker 테스트 코드 파일추가 및 구현

malrang-malrang committed:
✅ [test]: URLProductCatalogTests의 테스트 파일 생성 및 구현

malrang-malrang committed:
🎨[style]: 스타일 컨벤션 가이드를 준수하도록 코드 컨벤션 수정

malrang-malrang committed:
⚡️[improve]:URLSessionProvider session 프로퍼티 접근제어 추가

malrang-malrang committed:
♻️[refactor]: URL 관리 방법 변경

* enum Endpoint 를 활요하여 URL 을 관리하도록 변경하고 그에따른 로직 변경

Taeangel committed:
♻️[refactor]: 상황에 맞는 적절한 에러 변경

Taeangel committed:
🔥[del]: Decodable parse 메서드 삭제및 테스트코드 제거

malrang-malrang committed:
📝[docs]: STEP1 README Update

stevenkim18 commented on OpenMarket/OpenMarket/Model/ProductCatalog.swift:
```diff
@@ -0,0 +1,32 @@
+
+import Foundation
+
+struct ProductCatalog: Codable {
```
개인적인 의견이지만 Catalog 가 목록이라는 단어가 있긴 하지만..

`ProductList`가 조금 더 직관적인 네이밍이지 않을까 싶네요 ㅎㅎ 어떻게 생각하시나요? 아니면 더 좋은 이름이 있을까요?

malrang-malrang replied:
게속...productList 와 productCatalog 를 고민했습니다..!! 
아무래도 List 가 좀더 직관적일것 같습니다!! ㅎㅎ

stevenkim18 commented on OpenMarket/OpenMarket/Model/ProductCatalog.swift:
```diff
@@ -0,0 +1,32 @@
+
+import Foundation
+
+struct ProductCatalog: Codable {
```
이 모델이 Encoding(객체 -> json)도 하는 모델인가요? 

꼭 Codable을 채택해야 될까요?

malrang-malrang replied:
그렇군요! 디코딩만 가능하면되기 때문에 Decodable 을 채택하는것이 더 좋을것 같습니다!

stevenkim18 commented on OpenMarket/OpenMarket/Model/ProductCatalog.swift:
```diff
@@ -0,0 +1,32 @@
+    
+    enum CodingKeys: String, CodingKey {
+        case pageno = "page_no"
+        case itemspPerPage = "items_per_page"
```
오타가 있네요 ㅎㅎ

malrang-malrang replied:
자잘한 실수가 많았군요 ㅠㅠㅠ 꼼꼼히 검토하도록 하겠습니다!!🥲

stevenkim18 commented on OpenMarket/OpenMarket/Model/ProductCatalog.swift:
```diff
@@ -0,0 +1,32 @@
+        case offset
+        case limit
+        case pages
+        case lastPage = "last_page"
+        case nextPage = "has_next"
+        case prevPage = "has_prev"
```
lastpage가 정수라서 nextpage, prevpage도 정수형 처럼 느껴집니다.

원래 json 값 처럼 has가 들어가야 더 bool 타입 처럼 느껴지는데.. 어떻게 생각하시나요?

Taeangel replied:
```
    let hasNextPage: Bool?
    let hasPrevPage: Bool?

```
로 네이밍 변경하였습니다.

stevenkim18 commented on OpenMarket/OpenMarket/Model/Product.swift:
```diff
@@ -0,0 +1,35 @@
+//
+import Foundation
+
+struct Product: Codable {
+    let id: Int?
+    let vendorId: Int?
+    let name: String?
+    let thumbnail: String?
+    let currency: String?
+    let price: Int?
+    let bargainPrice: Int?
+    let discountedPrice: Int?
+    let stock: Int?
+    let createdAt: String?
+    let issuedAt: String?
+    
+    enum CodingKeys: String, CodingKey {
+        case id
+        case vendorId = "vendor_id"
+        case name
+        case thumbnail
+        case currency
+        case price
+        case bargainPrice = "bargain_price"
+        case discountedPrice = "discounted_price"
+        case stock
+        case createdAt = "created_at"
+        case issuedAt = "issued_at"
+    }
+}
```
codingKeys 사용하신 것은 매우 잘하셨습니다!!!

jsonDecoder 프로퍼티 중에서 스네이크케이스에서 카멜케이스로 바꿔주는 기능이 있습니다!! 한번 찾아보시겠어요?

그렇게 되면 enum 타입은 모두 지워도 제대로 동작합니다!
(엄청 중요한건 아니니 30분 이상 시간은 사용하지 마시고 못찾으시면 여쩌봐주세요!!)

malrang-malrang replied:
JsonDecoder 프로퍼티중 keyDecodingStrategy 요녀석이 있는것은 알고있었으나 사용해본적이 없었습니다!
이번 기회에 활용해보았는데 enum CodingKeys 를 제거해도 제대로 파싱 되는것을 확인했습니다! 😁
적용해서 수정하였습니다! :)

stevenkim18 commented on OpenMarket/OpenMarket/Network/Endpoint.swift:
```diff
@@ -0,0 +1,34 @@
+//
+//  API.swift
```
파일 이름과 타입이름이 다르네요!

Taeangel replied:
이름 수정했습니다!

stevenkim18 commented on OpenMarket/OpenMarket/Model/ProductCatalog.swift:
```diff
@@ -0,0 +1,32 @@
+//
+//  NetWork.swift
```
여기도 파일이름과 타입이름이 다르네요 ㅎㅎ

Taeangel replied:
ㅠㅠ 네

stevenkim18 commented on OpenMarket/OpenMarket/Model/Product.swift:
```diff
@@ -0,0 +1,35 @@
+//
+//  Products.swift
```
여기도.. ㅎㅎ

Taeangel replied:
ㅠㅠ네 수정했습니다

stevenkim18 commented on OpenMarket/OpenMarket/Network/URLSessionProtocol.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+typealias DataTaskCompletionHandler = (Data?, URLResponse?, Error?) -> Void
```
typealias를 사용하셨군요!! 잘하셨습니다!!!

stevenkim18 commented on OpenMarket/OpenMarket/Network/URLSessionProvider.swift:
```diff
@@ -0,0 +1,56 @@
+        self.session = session
+    }
+    
+    func getData(
```
Swift API 디자인 가이드에서 get은 어떨 때 함수명에 쓰일까요? 

조금 더 좋은 네이밍은 없을까요?

malrang-malrang replied:
get 메서드는 값을 얻어 올때 사용됩니다!
하지만 결과 값까지 반환되기 때문에 fatch 가 더 적절해보이기에 수정하였습니다! 😄

stevenkim18 commented on OpenMarket/OpenMarket/Network/URLSessionProvider.swift:
```diff
@@ -0,0 +1,56 @@
+    
+    func getData(
+        from url: Endpoint,
+        completionHandler: @escaping (Result<T, NetworkError>) -> Void
+    ) {
+        guard let url = url.url else {
+            completionHandler(.failure(.urlError))
+            return
+        }
+        
+        var request = URLRequest(url: url)
+        request.httpMethod = "GET"
+        
+        let task = session.dataTask(with: request) { data, urlResponse, error in
+            
+            guard error == nil else {
+                completionHandler(.failure(.clientError))
+                return
+            }
+            
+            guard let httpResponse = urlResponse as? HTTPURLResponse,
+                  (200...299).contains(httpResponse.statusCode) else {
+                completionHandler(.failure(.statusCodeError))
+                return
+            }
+            
+            guard let data = data else {
+                completionHandler(.failure(.dataError))
+                return
+            }
+            
+            guard let products = try? JSONDecoder().decode(T.self, from: data) else {
+                completionHandler(.failure(.decodeError))
+                return
+            }
+            
+            completionHandler(.success(products))
+        }
+        task.resume()
+    }
```
함수가 너무 깁니다. 

여러 함수로 나눠볼까요?

malrang-malrang replied:
네! 해당 함수 URLRequest 를 만들어주는 부분과 
디코딩 하는 부분을 따로 분리 하였습니다! 😁

stevenkim18 commented on OpenMarket/OpenMarket/Network/URLSessionProvider.swift:
```diff
@@ -0,0 +1,56 @@
+            guard let products = try? JSONDecoder().decode(T.self, from: data) else {
+                completionHandler(.failure(.decodeError))
+                return
+            }
```
guard으로 하신 것은 매우 보기 깔끔하네요  👍

stevenkim18 commented on OpenMarket/OpenMarket/URLHealthcheckerTests/URLHealthcheckerTests.swift:
```diff
@@ -0,0 +1,39 @@
+        }
+        wait(for: [promise], timeout: 10)
+    }
+}
```
실패하는 테스트 케이스는 없을까요?

stevenkim18 commented on OpenMarket/OpenMarket/URLProductCatalogTests/URLProductCatalogTests.swift:
```diff
@@ -0,0 +1,57 @@
+        sut = nil
+    }
+   
+    func test_URL이_상품리스트_조회_일때_fetchData_메서드를_호출하면_data의offset_값이_10인지() {
+        //given
+        let promise = expectation(description: "The offset value of data is 10")
+        
+        //when
+        sut.getData(from: .productList(page: 2, itemsPerPage: 10)) { result in
```
테스트 이름은 `fetchData`인데 함수이름은 `getData` 네요!

malrang-malrang replied:
앗..! 테스트 코드에서 변경하는것을 체크하지 못했습니다 ㅠ 
수정하였습니다! :)

stevenkim18 commented on OpenMarket/OpenMarket/URLProductTests/URLProductTests.swift:
```diff
@@ -0,0 +1,40 @@
+        sut = nil
+    }
+    
+    func test_URL이_api_products_2085_일때_fetchData_메서드를_호출하면_data의vendorId_값이_14인지() {
+        //given
+        let promise = expectation(description: "The vendorId value of data is 14")
+        
+        //when
+        sut.getData(from: .detailProduct(id: 2085)) { result in
+            //then
+            switch result {
+            case .success(let data):
+                XCTAssertEqual(data.vendorId, 14)
+            case .failure(_):
+                XCTFail()
+            }
+            promise.fulfill()
+        }
+        wait(for: [promise], timeout: 10)
+    }
+}
```
이 테스트 케이스만 보고 2085번을 요청하면 id가 14가 온다가 어떻게 성공하는 케이스 인줄 알 수 있나요?

테스트를 조금 더 일반화하거나 테스트 이름만 봐도 이제 어떤 테스트인지 대략적으로 감이 와야 합니다!!

Taeangel replied:
product의 값은 항상 같은 값을 테스트 할 수 없어 product 테스트는 삭제하였습니다.

stevenkim18 commented on OpenMarket/OpenMarket/URLProductTests/URLProductTests.swift:
```diff
@@ -0,0 +1,40 @@
+import XCTest
+@testable import OpenMarket
+
+class URLProductTests: XCTestCase {
+    
+    var sut: URLSessionProvider<Product>!
+    
+    override func setUpWithError() throws {
+        sut = URLSessionProvider()
+    }
+    
+    override func tearDownWithError() throws {
+        sut = nil
+    }
+    
+    func test_URL이_api_products_2085_일때_fetchData_메서드를_호출하면_data의vendorId_값이_14인지() {
+        //given
+        let promise = expectation(description: "The vendorId value of data is 14")
+        
+        //when
+        sut.getData(from: .detailProduct(id: 2085)) { result in
+            //then
+            switch result {
+            case .success(let data):
+                XCTAssertEqual(data.vendorId, 14)
+            case .failure(_):
+                XCTFail()
+            }
+            promise.fulfill()
+        }
+        wait(for: [promise], timeout: 10)
+    }
+}
```
그리고 Error 케이스가 6개인데 실패하는 테스트는 한개도 없네요 🥲

코드 커버리지를 측정해보지는 않았지만 아마 낮게 나올 것 같습니다.
실패하는 케이스도 만들어주세요!!

malrang-malrang replied:
해당 부분 MockURLSession 에 flag 를 두어 flag 가 true 라면 조건에 통과하는 satusCode 200 을 주고 false 라면 satusCode 400 을 주도록 변경하여 StatusCodeError 를 테스트하는 테스트 함수를 구현해두었습니다! 

조언해주셔 감사합니다 ! 😁

stevenkim18 replied:
URLProtocol 을 사용하면 모든 애러 케이스에 대한 테스트가 가능한 걸로 알고 있어요 ㅎㅎ

stevenkim18 commented:
태엔젤, 말랑 스탭1 하시느라 너무 고생많으셨어요!!
질문은 언제든지 환영입니다!!! 많이 질문해주시고 저한테 얻어갈 것들 많이 가져가세요!!

<br>

> 1. MockURLSessionDataTask init 관련 Error

네 맞습니다 ㅜㅜ `URLProtocol`을 활용해서 mock테스트 하는 것을 찾아보세요!!
<br>

>  2. 테스트를 하기위한 Mock 객체에서의 강제언래핑

강제 언레핑을 지양하기는 해야하지만 위에서 예외처리를 잘 해주었고 절대로 nil이 날 경우가 아니라면 저는 사용해도 괜찮다고 생각해요.
`urlRequest.url`이 언제 nil이 되는 경우를 알아보세요. 매개변수로 이상한 url만 넘겨주지 않는다면 앱이 크래쉬 날일은 없습니다.
제가 알기로는 빈 문자열을 URL에 넣으면 nil되는 걸로 알고 있습니다. `URLRequest` 에서 url을 빼는 것은 조금 더 찾아보셔야할 것 같아요

malrang-malrang committed:
♻️[refactor]: 오타제거 및 Codable에서  Decodable로 채택 변경

* Product , ProductList, URLSessionProvider 채택 변경

malrang-malrang committed:
⚡️[improve]: JsonDecoder keyDecodingStrategy 활용

*JsonDecoder keyDecodingStrategy convertFromSnakeCase 를 사용하여 파싱될 모델의 enum CodingKeys 제거

Taeangel committed:
🎨[style]: ProductCatalog 네이밍을 ProductList로 변경

Taeangel committed:
🎨[style]: 파일이름 주석 변경, 폴더 이름변경

Taeangel committed:
🎨[style]: Created by 수정

Taeangel committed:
🎨[style]: bool타입에 맞게 네이밍변경

malrang-malrang committed:
✨[feat]: Decodable+Extension 파일 추가및 구현

* parse() 메서드 구현

malrang-malrang committed:
♻️[refactor]: URLSessionProvider getData() 로직 수정

*네이밍 변경및 메서드 분리

malrang-malrang committed:
✏️[correct]: Test 코드의 getData() 메서드 네이밍 변경

* getData() -> fetchData() 로변경,
* 네이밍 변경에 따른 Test 메서드 네이밍 변경

Taeangel committed:
✏️[correct]: URLProductListTests로 네이밍 변경

Taeangel committed:
🔥[del]: URLProductTests 파일삭제

Taeangel committed:
✅[test]: statusCodeError받는부분 테스트 코드 작성

malrang-malrang committed:
🐛[fix]: MockURLSessionDataTask init 관련 Error 에러 해결

malrang-malrang committed:
✅[test]: statusCodeError Test, MockURLSession 수정

* MockURLSessionTest statusCodeError Test 추가
* MockURLSession isRequestSuccess 플래그 추가

stevenkim18 commented:
수정하신 코드 잘 봤어요!!! 크게 문제되는 점 없어서 머지하겠습니다!!

다음 스탭에서 collectionview로 화면 한개 구현해보시고 이후에 테스트 등에 대해서 더 고민해주세요!!

고생하셨습니다!!!
